lib/ddtrace/context.rb

```diff
@@ -0,0 +1,145 @@
+        # in other cases, this is just broken and one should rely
+        # on per-instrumentation code to retrieve handle parent/child relations.
+        @current_span = span.parent
+        return if span.tracer.nil?
+        return unless Datadog::Tracer.debug_logging
+        if span.parent.nil? && !check_finished_spans
+          opened_spans = @trace.length - @finished_spans
+          tracer.log.debug("root span #{span.name} closed but has #{opened_spans} unfinished spans:")
+          @trace.each do |s|
+            tracer.log.debug("unfinished span: #{s}") unless s.check_finished_spans
+          end
+        end
+      end
+    end
+
+    # Returns if the trace for the current Context is finished or not.
+    # Low-level internal function, not thread-safe.
+    def check_finished_spans
+      @finished_spans > 0 && @trace.length == @finished_spans
+    end
+
+    # Returns if the trace for the current Context is finished or not. A \Context
+    # is considered finished if all spans in this context are finished.
+    def finished?
+      @mutex.synchronize do
+        return check_finished_spans
+      end
+    end
+
+    # Returns true if the context is sampled, that is, if it should be kept
+    # and sent to the trace agent.
+    def sampled?
+      @mutex.synchronize do
+        return @sampled
+      end
+    end
+
+    # Returns both the trace list generated in the current context and
+    # if the context is sampled or not. It returns nil, nil if the ``Context`` is
+    # not finished. If a trace is returned, the \Context will be reset so that it
+    # can be re-used immediately.
+    #
+    # This operation is thread-safe.
+    def get
+      @mutex.synchronize do
+        return nil, nil unless check_finished_spans
+
+        trace = @trace
+        sampled = @sampled
+        reset
+        return trace, sampled
+      end
+    end
+
+    # Return a string representation of the context.
+    def to_s
+      @mutex.synchronize do
+        # rubocop:disable Metrics/LineLength
+        "Context(trace.length:#{@trace.length},sampled:#{@sampled},finished_spans:#{@finished_spans},current_span:#{@current_span})"
+      end
+    end
+
+    private :reset
+    private :check_finished_spans
+  end
+
+  # ThreadLocalContext can be used as a tracer global reference to create
+  # a different \Context for each thread. In synchronous tracer, this
+  # is required to prevent multiple threads sharing the same \Context
+  # in different executions.
+  class ThreadLocalContext
+    # ThreadLocalContext can be used as a tracer global reference to create
+    # a different \Context for each thread. In synchronous tracer, this
+    # is required to prevent multiple threads sharing the same \Context
+    # in different executions.
+    def initialize
+      self.local = Datadog::Context.new
+    end
+
+    # Override the thread-local context with a new context.
+    def local=(ctx)
```

Comment:
In Python this is a couple of `get/set` functions, I think in Ruby overloading `=()` is better for that setter/getter pattern. Additionally, while writing the code, I was confused by getting/setting too many things. At least this is different from `Context.get` (the difference between `get` on `ThreadLocalContext and` `get` on `Context` which returns a tuple `traces, sampled` was making things a little unclear).